Userland/Libraries/LibCore/MemoryStream.cpp

```diff
@@ -124,4 +125,123 @@ size_t FixedMemoryStream::remaining() const
     return m_bytes.size() - m_offset;
 }
 
+ErrorOr<Bytes> AllocatingMemoryStream::read(Bytes bytes)
+{
+    size_t read_bytes = 0;
+
+    while (read_bytes < bytes.size()) {
+        VERIFY(m_write_offset >= m_read_offset);
+
+        auto range = TRY(next_read_range());
+        if (range.size() == 0)
+            break;
+
+        auto copied_bytes = range.copy_trimmed_to(bytes.slice(read_bytes));
+
+        read_bytes += copied_bytes;
+        m_read_offset += copied_bytes;
+    }
+
+    cleanup_unused_chunks();
+
+    return bytes.trim(read_bytes);
+}
+
+ErrorOr<size_t> AllocatingMemoryStream::write(ReadonlyBytes bytes)
+{
+    size_t written_bytes = 0;
+
+    while (written_bytes < bytes.size()) {
+        VERIFY(m_write_offset >= m_read_offset);
+
+        auto range = TRY(next_write_range());
+
+        auto copied_bytes = bytes.slice(written_bytes).copy_trimmed_to(range);
+
+        written_bytes += copied_bytes;
+        m_write_offset += copied_bytes;
+    }
+
+    return written_bytes;
+}
+
+ErrorOr<void> AllocatingMemoryStream::discard(size_t count)
+{
+    VERIFY(m_write_offset >= m_read_offset);
+
+    if (count > size())
+        return Error::from_string_literal("Number of discarded bytes is higher than the number of allocated bytes");
+
+    m_read_offset += count;
+
+    cleanup_unused_chunks();
+
+    return {};
+}
+
+bool AllocatingMemoryStream::is_eof() const
+{
+    return size() == 0;
+}
+
+size_t AllocatingMemoryStream::size() const
+{
+    return m_write_offset - m_read_offset;
+}
```

Comment:
`SeekableStream::size` says:
> Returns the total size of the stream [...]

This seems to return how many bytes are available for reading instead. Can we make it so that this just returns `m_write_offset`? We don't need to make this a `SeekableStream` right now but it would still be great to keep POLA.

Reply:
I considered the size of "whatever we currently have in storage" to be the total size of the stream, especially since we deallocate the old chunks shortly after. That also makes the `copy_to_contiguous_buffer` replacement considerably easier.

Although, I agree, making this a SeekableStream (or SeekableStream-like) is a whole other :yaktangle:.

Reply:
Also, as already noted, I myself wasn't too sure whether making this an actual `Stream` is the perfect fit, but some decompressors and compressors use this as a drop-in replacement so that they can copy the whole thing into a buffer later, and I'm not looking to change that architecture as of now. That would only be possible once we took care of all the `compress_all` and `decompress_all` users.

Reply:
Of course, since we need to know that this is an `AllocatingMemoryStream` to use `size` anyways, we could just rename it to something that doesn't conflict with `SeekableStream` as much. Something like `used_buffer_size`?